### DISCUSSION TITLE
Use PyScript with jQuery

### DISCUSSION THREAD

**Discussion by wyatt-wong** [pyscript/pyscript#2221] [Q&A]
Is it possible to use PyScript with jQuery ?

**Comment by WebReflection** [comment]
```toml
[js_modules.main]
"https://esm.run/jquery" = "JQuery"
```

```python
from pyscript.js_modules.JQuery import default as jq

jq("#some-button-id").click();
```

See it live in here: https://pyscript.com/@agiammarchi/jquery-pyscript/v1?files=main.py

**Comment by WebReflection** [comment]
Apologies, I didn't mean to close this ... if the answer satisfied your question, please flag it as such, otherwise please explain more about your use case. The presented demo would work even from a worker, but in there jQuery would be slower due roundtrips needed to chain references from the *main* thread ... yet it should work without any issue keeping the config as it is.

**Reply by wyatt-wong** [reply]
I guessed your example is over simplified. I would like to know more:

1. Do I need to add jQuery in the index.html such as the following?

`<script
  src="https://code.jquery.com/jquery-3.7.1.slim.js"
  integrity="sha256-UgvvN8vBkgO0luPSUl2s8TIlOSYRoGFAX4jlCIm9Adc="
  crossorigin="anonymous"></script>`

2. If I put jQuery functions in a separate javascript file, do I need to write this ?

`$(document).ready(function(){

  // jQuery methods go here...

});`

I asked this because I saw the PyScript example use DOM to access the elements in the HTML document but I have not used the DOM style syntax for a long time. Most people use jQuery to simplify the coding in JavaScript or even TypeScript. So I wonder if it is possible to mix jQuery with PyScript to simplify the coding process.

**Reply by WebReflection** [reply]
> Do I need to add jQuery in the index.html such as the following?

no, you import jQuery as a module and that's it, no global pollution needed

> If I put jQuery functions in a separate javascript file, do I need to write this ?

from Python you import jQuery as module as shown in my example

> Most people use jQuery to simplify the coding in JavaScript or even TypeScript. So I wonder if it is possible to mix jQuery with PyScript to simplify the coding process.

I might not understand your question then ... do you need jQuery in Python or you just want to add jQuery with JavaScript?

In both cases using a module is more 2024 and less 2000 but if you have other JS scripts that relies on the fact a global `$` reference exists, then your script with integrity on the page would be the old way to go.

**Reply by wyatt-wong** [reply]
I just found out PyQuery for Python so it looks like I don't need to directly embed jQuery javascript into the HTML documents.

[PyQuery](https://pyquery.readthedocs.io/l)

**Reply by WebReflection** [reply]
my first example literally gives you *jQuery* as it is in Python ... you never need to directly embed jQuery into the HTML document, my example shows that via a `config.toml`, no jQuery on the page, not on the HTML, not on the dom, nowhere ... it's a JS module imported into Python. I think I've done a very bad job at explaining what is that example, it's **100% jQuery** on Python, no indirection whatsoever.

**Reply by wyatt-wong** [reply]
I followed your example but did it in VS Code in my local environment. I got warning in the main.py which cannot resolve the "pyscript.js_modules.JQuery" even though I setup the pyscript.toml exactly like yours.

![Screenshot 2024-10-17 at 2 51 43 AM](https://github.com/user-attachments/assets/b22baa41-b28f-4a33-9d15-e8a44c3c9efc)

I have already included the stylesheet and module under <head></head> and also included following statement in the index.html

<script type="py" src="./main.py" config="./pyscript.toml"></script>

**Reply by WebReflection** [reply]
what you see on PyScript .com is the current PyScript version importing jQuery and there's nothing special in there. If you can't import from pyscript.js_modules it's very likely you are using a very outdated version of pyscript.

The live example that works in PyScript .com has an index.html file that shows you how to use latest PyScript in your project too.

There is no "work on my machine" here because we have tons of tests around modules and we use modules ourselves so I'd be very surprised to learn you doing exactly the same would not work in there ... unless you are running code in a worker and you forgot to enable special headers to do so, all explained in our documentation though.

Can you please show me your HTML page or how are you bootstrapping PyScript in general?

**Comment by Neon22** [comment]
Its also worth noting that the ltk module uses JQuery. (3.7)
the ltk provides very good (IMHO) dom access and UI control.

So you get JQuery fully integrated as a side effect. All the work has been done
